File: api_python/examples/100-Overview/02-protection_zones_configuration.py
#! /usr/bin/env python3

###
# KINOVA (R) KORTEX (TM)
#
# Copyright (c) 2019 Kinova inc. All rights reserved.
#
# This software may be modified and distributed
# under the terms of the BSD 3-Clause license.
#
# Refer to the LICENSE file for details.
#
###

import sys
import os
import time
import threading

from kortex_api.autogen.client_stubs.BaseClientRpc import BaseClient

from kortex_api.autogen.messages import Base_pb2, Common_pb2

# Position of the protection zone (in meters)
PROTECTION_ZONE_POS =  [0.75, 0.0, 0.4]

# Size of the protection zone (in meters)
PROTECTION_ZONE_DIMENSIONS = [0.05, 0.3, 0.4]

# Theta values of the protection zone movement (in degrees)
PROTECTION_ZONE_MOVEMENT_THETAS = [90.0, 0.0, 90.0]

# Waiting time between actions (in milliseconds)
ACTION_WAITING_TIME = 1

# Create closure to set finished to true after an END or an ABORT
def check_for_end_or_abort(e):
    """Return a closure checking for END or ABORT notifications

    Arguments:
    finished -- list of bool to affect the result (of size 1)
                (will be set to True when an END or ABORT occurs)
    """
    def check(notification, e = e):
        print("EVENT : " + \
              Base_pb2.ActionEvent.Name(notification.action_event))
        if notification.action_event == Base_pb2.ACTION_END \
        or notification.action_event == Base_pb2.ACTION_ABORT:
            e.set()
    return check
 
def move_to_home_position(base):
    # Make sure the arm is in Single Level Servoing mode
    base_servo_mode = Base_pb2.ServoingModeInformation()
    base_servo_mode.servoing_mode = Base_pb2.SINGLE_LEVEL_SERVOING
    base.SetServoingMode(base_servo_mode)
    
    # Move arm to ready position
    print("Moving the arm to a safe position")
    action_type = Base_pb2.RequestedActionType()
    action_type.action_type = Base_pb2.REACH_JOINT_ANGLES
    action_list = base.ReadAllActions(action_type)
    action_handle = None
    for action in action_list.action_list:
        if action.name == "Home":
            action_handle = action.handle

    if action_handle == None:
        print("Can't reach safe position. Exiting")
        sys.exit(1)

    e = threading.Event()
    notification_handle = base.OnNotificationActionTopic(
        check_for_end_or_abort(e),
        Base_pb2.NotificationOptions()
    )

    base.ExecuteActionFromReference(action_handle)

    e.wait()

    base.Unsubscribe(notification_handle)

def move_to_protection_zone(base):
    
    command = Base_pb2.TwistCommand()

    command.reference_frame = Base_pb2.CARTESIAN_REFERENCE_FRAME_BASE
    command.duration = 0

    twist = command.twist
    twist.linear_x = 0.05
    twist.linear_y = 0
    twist.linear_z = 0
    twist.angular_x = 0
    twist.angular_y = 0
    twist.angular_z = 0

    print ("Moving towards the protection zone for 4 seconds...")
    base.SendTwistCommand(command)

    # Let time for twist to be executed
    time.sleep(4)

    print ("Stopping the robot...")
    base.Stop()
    time.sleep(1)

def move_in_front_of_protection_zone(base):
    
    print("Starting Cartesian action movement ...")
    action = Base_pb2.Action()
    action.name = "Example Cartesian action movement"
    action.application_data = ""
 
    cartesian_pose = action.reach_pose.target_pose
    cartesian_pose.x = PROTECTION_ZONE_POS[0] - 0.1     # (meters)
    cartesian_pose.y = PROTECTION_ZONE_POS[1]           # (meters)
    cartesian_pose.z = PROTECTION_ZONE_POS[2]           # (meters)
    cartesian_pose.theta_x = PROTECTION_ZONE_MOVEMENT_THETAS[0] # (degrees)
    cartesian_pose.theta_y = PROTECTION_ZONE_MOVEMENT_THETAS[1] # (degrees)
    cartesian_pose.theta_z = PROTECTION_ZONE_MOVEMENT_THETAS[2] # (degrees)

    e = threading.Event()
    notification_handle = base.OnNotificationActionTopic(
        check_for_end_or_abort(e),
        Base_pb2.NotificationOptions()
    )

    print("Executing action")
    base.ExecuteAction(action)

    print("Waiting for movement to finish ...")
    e.wait()

    print("Cartesian movement completed")

    base.Unsubscribe(notification_handle)

def print_protection_zones(base):

    all_protection_zones = base.ReadAllProtectionZones()

    print("PROTECTION ZONES")
    for protection_zone in all_protection_zones.protection_zones:
        message = "Protection Zone : " + protection_zone.name + \
        " Origin : [ " \
        + str(protection_zone.shape.origin.x) + " "\
        + str(protection_zone.shape.origin.y) + " "\
        + str(protection_zone.shape.origin.z) \
        + " ] Dimensions : [ "
        for dim in protection_zone.shape.dimensions:
            message += str(dim) + " "
        message += "]"
        print(message)

def create_protection_zone(base):

    zone = Base_pb2.ProtectionZone()

    zone.name = "Example Protection Zone"
    zone.is_enabled = True
    shape = zone.shape
    shape.shape_type = Base_pb2.RECTANGULAR_PRISM

    point = shape.origin
    point.x = PROTECTION_ZONE_POS[0]
    point.y = PROTECTION_ZONE_POS[1]
    point.z = PROTECTION_ZONE_POS[2]
    shape.dimensions.append(PROTECTION_ZONE_DIMENSIONS[0])
    shape.dimensions.append(PROTECTION_ZONE_DIMENSIONS[1])
    shape.dimensions.append(PROTECTION_ZONE_DIMENSIONS[2])

    shape.orientation.row1.column1 = 1.0
    shape.orientation.row2.column2 = 1.0
    shape.orientation.row3.column3 = 1.0

    return base.CreateProtectionZone(zone)

def main():
    
    # Import the utilities helper module
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
    import utilities

    # Parse arguments
    args = utilities.parseConnectionArguments()
    
    # Create connection to the device and get the router
    with utilities.DeviceConnection.createTcpConnection(args) as router:

        # Create required services
        base = BaseClient(router)

        # Example core

        # Move to initial position
        move_to_home_position(base)

        # Move without the protection zone
        print_protection_zones(base)
        move_in_front_of_protection_zone(base)
        move_to_protection_zone(base)
        move_to_home_position(base)

        # Move with the protection zone
        print_protection_zones(base)
        move_in_front_of_protection_zone(base)

        # Add the protection zone
        handle = create_protection_zone(base)

        move_to_protection_zone(base)
        move_to_home_position(base)

        # Delete the protection zone
        base.DeleteProtectionZone(handle)

        # Print final protection zones
        # The example protection zone should be removed
        print_protection_zones(base)

if __name__ == "__main__":
    main()

File: api_cpp/examples/100-Overview/02-protection_zones_configuration.cpp
/*
* KINOVA (R) KORTEX (TM)
*
* Copyright (c) 2019 Kinova inc. All rights reserved.
*
* This software may be modified and distributed
* under the terms of the BSD 3-Clause license.
*
* Refer to the LICENSE file for details.
*
*/

#include <BaseClientRpc.h>
#include <SessionManager.h>

#include <RouterClient.h>
#include <TransportClientTcp.h>

#include <array>

#include "utilities.h"

#define PORT 10000

namespace k_api = Kinova::Api;

// Position of the protection zone (in meters)
const std::array<float,3> PROTECTION_ZONE_POS
{
    0.75f, 
    0.00f, 
    0.40f
};

// Size of the protection zone (in meters)
const std::array<float,3> PROTECTION_ZONE_DIMENSIONS
{
    0.05f, 
    0.30f, 
    0.40f
};

/// Theta values of the protection zone movement (in degrees)
const std::array<float,3> PROTECTION_ZONE_MOVEMENT_THETAS
{
    90.0,
    00.0,
    90.0 
};

// Waiting time during actions
const auto ACTION_WAITING_TIME = std::chrono::seconds(1);

// Create closure to set finished to true after an END or an ABORT
std::function<void(k_api::Base::ActionNotification)> 
check_for_end_or_abort(bool& finished)
{
    return [&finished](k_api::Base::ActionNotification notification)
    {
        std::cout << "EVENT : " << k_api::Base::ActionEvent_Name(notification.action_event()) << std::endl;

        // The action is finished when we receive a END or ABORT event
        switch(notification.action_event())
        {
        case k_api::Base::ActionEvent::ACTION_ABORT:
        case k_api::Base::ActionEvent::ACTION_END:
            finished = true;
            break;
        default:
            break;
        }
    };
}

bool move_to_home_position(k_api::Base::BaseClient* base)
{
    // Make sure the arm is in Single Level Servoing before executing an Action
    auto servoingMode = k_api::Base::ServoingModeInformation();
    servoingMode.set_servoing_mode(k_api::Base::ServoingMode::SINGLE_LEVEL_SERVOING);
    base->SetServoingMode(servoingMode);
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    // Move arm to ready position
    std::cout << "Moving the arm to a safe position" << std::endl;
    auto action_type = k_api::Base::RequestedActionType();
    action_type.set_action_type(k_api::Base::REACH_JOINT_ANGLES);
    auto action_list = base->ReadAllActions(action_type);
    auto action_handle = k_api::Base::ActionHandle();
    action_handle.set_identifier(0);
    for (auto action : action_list.action_list()) 
    {
        if (action.name() == "Home") 
        {
            action_handle = action.handle();
        }
    }

    if (action_handle.identifier() == 0) 
    {
        std::cout << "Can't reach safe position, exiting" << std::endl;
        return false;
    } 
    else 
    {
        bool action_finished = false; 
        // Notify of any action topic event
        auto options = k_api::Common::NotificationOptions();
        auto notification_handle = base->OnNotificationActionTopic(
            check_for_end_or_abort(action_finished),
            options
        );
        base->ExecuteActionFromReference(action_handle);

        // Leave time for action to finish
        while(!action_finished)
        { 
            std::this_thread::sleep_for(ACTION_WAITING_TIME);
        }

        base->Unsubscribe(notification_handle);
    }
    return true;
}

void move_to_protection_zone(k_api::Base::BaseClient* base)
{
    auto command = k_api::Base::TwistCommand();
    command.set_reference_frame(k_api::Common::CARTESIAN_REFERENCE_FRAME_BASE);
    command.set_duration(0);  // Unlimited time to execute

    std::cout << "Moving towards the protection zone for 4 seconds..." << std::endl;

    auto twist = command.mutable_twist();
    twist->set_linear_x(0.05f);
    twist->set_linear_y(0.0f);
    twist->set_linear_z(0.0f);
    twist->set_angular_x(0.0f);
    twist->set_angular_y(0.0f);
    twist->set_angular_z(0.0f);
    base->SendTwistCommand(command);

    // Let time for twist to be executed
    std::this_thread::sleep_for(std::chrono::milliseconds(4000));

    // Make movement stop
    base->Stop();
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
}

void move_in_front_of_protection_zone(k_api::Base::BaseClient* base)
{
    auto constrained_pose = k_api::Base::ConstrainedPose();
    auto pose = constrained_pose.mutable_target_pose();

    pose->set_x(PROTECTION_ZONE_POS.at(0) - 0.1f);    // x (meters)
    pose->set_y(PROTECTION_ZONE_POS.at(1));           // y (meters)
    pose->set_z(PROTECTION_ZONE_POS.at(2));           // z (meters)
    pose->set_theta_x(PROTECTION_ZONE_MOVEMENT_THETAS.at(0)); // theta x (degrees)
    pose->set_theta_y(PROTECTION_ZONE_MOVEMENT_THETAS.at(1)); // theta y (degrees)
    pose->set_theta_z(PROTECTION_ZONE_MOVEMENT_THETAS.at(2)); // theta z (degrees)

    bool action_finished = false; 
    // Notify of any action topic event
    auto options = k_api::Common::NotificationOptions();
    auto notification_handle = base->OnNotificationActionTopic(
        check_for_end_or_abort(action_finished),
        options
    );

    std::cout << "Reaching cartesian pose..." << std::endl;
    base->PlayCartesianTrajectory(constrained_pose);

    std::cout << "Waiting  for movement to finish ..." << std::endl;
    while(!action_finished)
    {
        std::this_thread::sleep_for(ACTION_WAITING_TIME);
    }

    std::cout << "Cartesian pose reached" << std::endl;

    base->Unsubscribe(notification_handle);
}

k_api::Base::ProtectionZoneHandle create_protection_zone(k_api::Base::BaseClient* base)
{
    // Create zone
    k_api::Base::ProtectionZone zone;
    zone.set_name("Example Protection Zone ");
    zone.set_is_enabled(true);

    // Set shape
    auto shape = zone.mutable_shape();
    shape->set_shape_type(k_api::Base::ShapeType::RECTANGULAR_PRISM);
    auto point = shape->mutable_origin();
    point->set_x(PROTECTION_ZONE_POS.at(0));
    point->set_y(PROTECTION_ZONE_POS.at(1));
    point->set_z(PROTECTION_ZONE_POS.at(2));
    shape->add_dimensions(PROTECTION_ZONE_DIMENSIONS.at(0));
    shape->add_dimensions(PROTECTION_ZONE_DIMENSIONS.at(1));
    shape->add_dimensions(PROTECTION_ZONE_DIMENSIONS.at(2));

    // Set orientation to Identity matrix
    auto orientation = shape->mutable_orientation();
    orientation->mutable_row1()->set_column1(1.0f);
    orientation->mutable_row2()->set_column2(1.0f);
    orientation->mutable_row3()->set_column3(1.0f);

    return base->CreateProtectionZone(zone);
}

void print_protection_zones(k_api::Base::BaseClient* base)
{
    const auto all_protection_zones = base->ReadAllProtectionZones();

    std::cout << "PROTECTION ZONES : " << std::endl;
    for(auto protection_zone : all_protection_zones.protection_zones())
    {
        std::cout 
            << "Protection Zone : " << protection_zone.name()
            << " Origin : [ " 
            << protection_zone.shape().origin().x() << " "
            << protection_zone.shape().origin().y() << " "
            << protection_zone.shape().origin().z()
            << " ] Dimensions : [ ";
            for(auto dim : protection_zone.shape().dimensions())
            {
                std::cout << dim << " ";
            }
            std::cout << "]" << std::endl;
    }
}

int main(int argc, char **argv)
{
    auto parsed_args = ParseExampleArguments(argc, argv);

    // Create API objects
    auto error_callback = [](k_api::KError err){ cout << "_________ callback error _________" << err.toString(); };
    auto transport = new k_api::TransportClientTcp();
    auto router = new k_api::RouterClient(transport, error_callback);
    transport->connect(parsed_args.ip_address, PORT);

    // Set session data connection information
    auto create_session_info = k_api::Session::CreateSessionInfo();
    create_session_info.set_username(parsed_args.username);
    create_session_info.set_password(parsed_args.password);
    create_session_info.set_session_inactivity_timeout(60000);   // (milliseconds)
    create_session_info.set_connection_inactivity_timeout(2000); // (milliseconds)

    // Session manager service wrapper
    std::cout << "Creating session for communication" << std::endl;
    auto session_manager = new k_api::SessionManager(router);
    session_manager->CreateSession(create_session_info);
    std::cout << "Session created" << std::endl;

    // Create services
    auto base = new k_api::Base::BaseClient(router);

    // Example core

    // Try to Move to initial position
    if(!move_to_home_position(base))
    {
        std::cout << "Home position not available. Stopping example..." << std::endl;
        return 1;
    }

    // Move without the protection zone
    print_protection_zones(base);
    std::cout << "Moving in front of protection zone..." << std::endl;
    move_in_front_of_protection_zone(base);
    std::cout << "Moving inside disabled protection zone..." << std::endl;
    move_to_protection_zone(base);
    std::cout << "Going back to home position..." << std::endl;
    move_to_home_position(base);

    // Add the protection zone
    std::cout << "Creating and enabling protection zone..." << std::endl;
    auto protection_zone_handle = create_protection_zone(base);

    // Move with the protection zone
    print_protection_zones(base);
    std::cout << "Moving in front of protection zone..." << std::endl;
    move_in_front_of_protection_zone(base);
    std::cout << "Moving inside enabled protection zone..." << std::endl;
    move_to_protection_zone(base);
    std::cout << "Going back to home position..." << std::endl;
    move_to_home_position(base);

    // Delete the protection zone
    base->DeleteProtectionZone(protection_zone_handle);

    // Print final protection zones
    // The example protection zone should be removed
    print_protection_zones(base);

    // Close API session
    session_manager->CloseSession();

    // Deactivate the router and cleanly disconnect from the transport object
    router->SetActivationStatus(false);
    transport->disconnect();

    // Destroy the API
    delete base;
    delete session_manager;
    delete router;
    delete transport;
}